src/encoder/writer.rs

```diff
@@ -20,83 +19,73 @@ impl TiffByteOrder for BigEndian {
     fn write_header<W: Write>(writer: &mut TiffWriter<W>) -> TiffResult<()> {
         writer.writer.write_u16::<BigEndian>(0x4d4d)?;
         writer.writer.write_u16::<BigEndian>(42)?;
-        writer.offset += 4;
 
         Ok(())
     }
 }
 
 pub struct TiffWriter<W> {
     writer: W,
-    offset: u64,
 }
 
 impl<W: Write> TiffWriter<W> {
     pub fn new(writer: W) -> Self {
-        Self { writer, offset: 0 }
-    }
-
-    pub fn offset(&self) -> u64 {
-        self.offset
+        Self { writer }
     }
 
     pub fn write_bytes(&mut self, bytes: &[u8]) -> Result<(), io::Error> {
         self.writer.write_all(bytes)?;
-        self.offset += bytes.len() as u64;
         Ok(())
     }
 
     pub fn write_u8(&mut self, n: u8) -> Result<(), io::Error> {
         self.writer.write_u8(n)?;
-        self.offset += 1;
         Ok(())
     }
 
     pub fn write_i8(&mut self, n: i8) -> Result<(), io::Error> {
         self.writer.write_i8(n)?;
-        self.offset += 1;
         Ok(())
     }
 
     pub fn write_u16(&mut self, n: u16) -> Result<(), io::Error> {
         self.writer.write_u16::<NativeEndian>(n)?;
-        self.offset += 2;
 
         Ok(())
     }
 
     pub fn write_i16(&mut self, n: i16) -> Result<(), io::Error> {
         self.writer.write_i16::<NativeEndian>(n)?;
-        self.offset += 2;
 
         Ok(())
     }
 
     pub fn write_u32(&mut self, n: u32) -> Result<(), io::Error> {
         self.writer.write_u32::<NativeEndian>(n)?;
-        self.offset += 4;
 
         Ok(())
     }
 
     pub fn write_i32(&mut self, n: i32) -> Result<(), io::Error> {
         self.writer.write_i32::<NativeEndian>(n)?;
-        self.offset += 4;
 
         Ok(())
     }
 
     pub fn write_u64(&mut self, n: u64) -> Result<(), io::Error> {
         self.writer.write_u64::<NativeEndian>(n)?;
-        self.offset += 8;
 
         Ok(())
     }
 
+}
+
+impl<W: Write + Seek> TiffWriter<W> {
     pub fn pad_word_boundary(&mut self) -> Result<(), io::Error> {
-        if self.offset % 4 != 0 {
+        let offset = self.offset();
+        if offset % 4 != 0 {
```

Comment:
This is not equivalent to the previous code as it nows aligns to the offset of the underlying writer instead of the offset within the tiff. Maybe this is connected to the cause of the issue? I do not understand enough about this specific part of the Tiff specification to say but your PR comment does not reference this in as a root cause.

Reply:
I think this is more correct. When I think about it, the previous code would have a problem when the writer does not start on offset 0. This is because the code sets the starting offset to 0 and then uses `seek(SeekFrom::Start())` which would then be incorrect. This should be fixable by changing the constructor to check the offset or use `SeekFrom::Current` instead. Not sure if it would fix this issue though.

Reply:
The previous code was mistaken in its implementation of `goto_offset`. But the scope of the PR is way larger than that and does not only address that. And I fear that this also introduced *another* subtle bug here despite trying to correct seeking issues.

Reply:
Specifically, the root cause seems to be located here https://github.com/image-rs/image-tiff/blob/6aea7b33c1d4c0b0c7384588513313ae9a52ae59/src/encoder/writer.rs#L108-L111

Note that `offset` should be relative to some stored starting offset. This PR abandons all of the offset tracking for fallible I/O operations (which are more expensive sys calls depending on the underlying writer) without sufficient motivation. And these changes in total affect more than the bug it lists. Is there some other bug I haven't seen? It does not explain.

Reply:
I also agree that the manual offset tracking is better, which is the reason I wrote it that way to begin with. 

This PR also adds some changes to the IFD offset writing. I believe (though not entirely convinced) that these changes are correct. The problem that these changes address is what causes problems with multi-image tiff.